File: apps/android/app/src/main/kotlin/app/pbbls/android/services/ReferenceDataService.kt
package app.pbbls.android.services

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.staticCompositionLocalOf
import app.pbbls.android.features.path.models.Domain
import app.pbbls.android.features.path.models.PebbleCollection
import app.pbbls.android.features.profile.models.SoulRow
import app.pbbls.android.features.profile.models.SoulWithGlyph
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

/**
 * Session cache of the three reference lists the pebble form needs (domains,
 * souls, collections) — ports ReferenceDataService.swift. `load()` is kicked from
 * RootScreen once a user id is present (souls/collections are RLS-scoped). No
 * retry: a failed load leaves empty lists (empty pickers), recovers next launch.
 * `refreshSouls()` closes iOS's stale-list gap after inline soul creation (D11).
 */
class ReferenceDataService(
    private val supabase: SupabaseService,
) {
    var domains: List<Domain> by mutableStateOf(emptyList())
        private set

    var souls: List<SoulWithGlyph> by mutableStateOf(emptyList())
        private set

    var collections: List<PebbleCollection> by mutableStateOf(emptyList())
        private set

    var hasLoaded: Boolean by mutableStateOf(false)
        private set

    suspend fun load() {
        try {
            coroutineScope {
                val domainsDeferred = async { fetchDomains() }
                val soulsDeferred = async { fetchSouls() }
                val collectionsDeferred = async { fetchCollections() }
                domains = domainsDeferred.await()
                souls = soulsDeferred.await()
                collections = collectionsDeferred.await()
                hasLoaded = true
            }
            Log.i(TAG, "loaded ${domains.size} domains, ${souls.size} souls, ${collections.size} collections")
        } catch (e: Exception) {
            Log.e(TAG, "reference data load failed — pickers empty until next launch", e)
        }
    }

    suspend fun refreshSouls() {
        try {
            souls = fetchSouls()
        } catch (e: Exception) {
            Log.e(TAG, "souls refresh failed", e)
        }
    }

    suspend fun refreshCollections() {
        try {
            collections = fetchCollections()
        } catch (e: Exception) {
            Log.e(TAG, "collections refresh failed", e)
        }
    }

    private suspend fun fetchDomains(): List<Domain> =
        supabase.client
            .from("domains")
            .select {
                order("name", Order.ASCENDING)
            }.decodeList<Domain>()

    private suspend fun fetchSouls(): List<SoulWithGlyph> =
        supabase.client
            .from("souls")
            .select(Columns.raw("id, name, glyph_id, glyphs(id, name, strokes, view_box)")) {
                order("name", Order.ASCENDING)
            }.decodeList<SoulRow>()
            .map { it.toSoulWithGlyph() }

    private suspend fun fetchCollections(): List<PebbleCollection> =
        supabase.client
            .from("collections")
            .select(Columns.list("id", "name")) {
                order("name", Order.ASCENDING)
            }.decodeList<PebbleCollection>()

    companion object {
        private const val TAG = "reference-data"
    }
}

val LocalReferenceDataService =
    staticCompositionLocalOf<ReferenceDataService> {
        error("LocalReferenceDataService not provided — wrap the tree in MainActivity's CompositionLocalProvider")
    }

File: apps/ios/Pebbles/Services/ReferenceDataService.swift
import Foundation
import Observation
import Supabase
import os

/// Caches the three reference lists used by `CreatePebbleSheet` and
/// `EditPebbleSheet` (`domains`, `souls`, `collections`) for the session.
///
/// Loaded once from `RootView.task` alongside `EmotionPaletteService`, during
/// the handcrafted logo loader; the loader gates on `didFinishLoading`.
/// Subsequent sheet opens read directly from the cached arrays — no per-open
/// round-trips. `domains` is seed data and never refreshes at runtime;
/// `souls` and `collections` are refreshed via `refreshSouls()` /
/// `refreshCollections()` after the matching Profile mutations succeed.
///
/// No retry on failure: a failed `load()` leaves the arrays empty and pickers
/// render empty, matching the UX of an offline launch. State recovers on the
/// next app launch.
@Observable
@MainActor
final class ReferenceDataService {
    private(set) var domains: [Domain] = []
    private(set) var souls: [SoulWithGlyph] = []
    private(set) var collections: [PebbleCollection] = []
    private(set) var hasLoaded: Bool = false
    /// True once a load attempt has settled — success OR failure. The launch
    /// loader gates on this (not `hasLoaded`) so a failed reference fetch still
    /// lets the app open with an empty cache instead of boiling forever.
    private(set) var didFinishLoading: Bool = false

    private let client: SupabaseClient
    private let logger = Logger(subsystem: "app.pbbls.ios", category: "reference-data")

    init(client: SupabaseClient) {
        self.client = client
    }

    /// Fetch all three lists in parallel and populate the cache. Idempotent —
    /// the splash-driven call site only fires once, but safe to call again
    /// (e.g. for retry after a transient launch-time failure).
    func load() async {
        defer { didFinishLoading = true }
        do {
            async let domainsQuery: [Domain] = client
                .from("domains")
                .select()
                .order("name")
                .execute()
                .value
            async let soulsQuery: [SoulWithGlyph] = client
                .from("souls")
                .select("id, name, glyph_id, glyphs(id, name, strokes, view_box)")
                .order("name")
                .execute()
                .value
            async let collectionsQuery: [PebbleCollection] = client
                .from("collections")
                .select("id, name")
                .order("name")
                .execute()
                .value

            let (loadedDomains, loadedSouls, loadedCollections) =
                try await (domainsQuery, soulsQuery, collectionsQuery)

            self.domains = loadedDomains
            self.souls = loadedSouls
            self.collections = loadedCollections
            self.hasLoaded = true
            logger.info("""
                loaded \(loadedDomains.count, privacy: .public) domains, \
                \(loadedSouls.count, privacy: .public) souls, \
                \(loadedCollections.count, privacy: .public) collections
                """)
        } catch {
            logger.error("reference data load failed: \(error.localizedDescription, privacy: .private)")
        }
    }

    /// Re-fetch souls only. Called from `SoulsListView` after create/edit/delete
    /// completes, so the pebble sheets see the new list on next open.
    func refreshSouls() async {
        do {
            let rows: [SoulWithGlyph] = try await client
                .from("souls")
                .select("id, name, glyph_id, glyphs(id, name, strokes, view_box)")
                .order("name")
                .execute()
                .value
            self.souls = rows
        } catch {
            logger.error("souls refresh failed: \(error.localizedDescription, privacy: .private)")
        }
    }

    /// Re-fetch collections only. Called from `CollectionsListView` after
    /// create/edit/delete completes.
    func refreshCollections() async {
        do {
            let rows: [PebbleCollection] = try await client
                .from("collections")
                .select("id, name")
                .order("name")
                .execute()
                .value
            self.collections = rows
        } catch {
            logger.error("collections refresh failed: \(error.localizedDescription, privacy: .private)")
        }
    }
}
